File: app/src/main/java/com/waveneuro/ui/dashboard/edit_client/EditClientBottomSheet.java
package com.waveneuro.ui.dashboard.edit_client;

import static com.waveneuro.utils.DateUtil.PATTERN_RFC1123;

import android.os.Bundle;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RadioButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.datepicker.MaterialDatePicker;
import com.google.android.material.textfield.TextInputEditText;
import com.waveneuro.R;
import com.waveneuro.domain.usecase.patient.GetPatientUseCase;
import com.waveneuro.domain.usecase.patient.UpdatePatientUseCase;
import com.waveneuro.injection.component.DaggerFragmentComponent;
import com.waveneuro.injection.component.FragmentComponent;
import com.waveneuro.injection.module.FragmentModule;
import com.waveneuro.ui.base.BaseActivity;
import com.waveneuro.ui.dashboard.home.HomeClientsViewState;
import com.waveneuro.ui.dashboard.home.HomeFragment;
import com.waveneuro.ui.dashboard.home.HomeUserViewState;
import com.waveneuro.utils.DateUtil;

import java.util.Date;

import javax.inject.Inject;

import butterknife.OnClick;

public class EditClientBottomSheet extends BottomSheetDialogFragment {

    protected FragmentComponent fragmentComponent;

    private int id;
    private String firstName;
    private String lastName;
    private String dob;
    private boolean isMale;
    private String email;
    private String username;
    private String organization;
    private boolean tos;

    TextInputEditText etFirstName;
    TextInputEditText etLastName;
    TextInputEditText etBirthdate;
    TextInputEditText etEmail;

    RadioButton rbMale;
    RadioButton rbFemale;

    @Inject
    EditClientViewModel editClientViewModel;

    EditClientViewModel.OnClientUpdated listener;


    public static EditClientBottomSheet newInstance(EditClientViewModel.OnClientUpdated listener, int id, String name, String lastName, String dob, boolean sex, String email, String username, String organization, boolean tos) {
        EditClientBottomSheet editClientBottomSheet = new EditClientBottomSheet();
        editClientBottomSheet.id = id;
        editClientBottomSheet.firstName = name;
        editClientBottomSheet.lastName = lastName;
        editClientBottomSheet.dob = dob;
        editClientBottomSheet.isMale = sex;
        editClientBottomSheet.email = email;
        editClientBottomSheet.username = username;
        editClientBottomSheet.organization = organization;
        editClientBottomSheet.tos = tos;
        editClientBottomSheet.listener = listener;
        return editClientBottomSheet;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        fragmentComponent = DaggerFragmentComponent.builder()
                .activityComponent(((BaseActivity) getActivity()).activityComponent())
                .fragmentModule(new FragmentModule(this))
                .build();

        fragmentComponent.inject(this);
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.dialog_edit_client, container,
                false);
        etFirstName = view.findViewById(R.id.tip_firstName);
        etLastName = view.findViewById(R.id.tip_lastName);
        etBirthdate = view.findViewById(R.id.tip_birthday);
        etEmail = view.findViewById(R.id.tip_email);
        rbMale = view.findViewById(R.id.rb_Male);
        rbFemale = view.findViewById(R.id.rb_Female);

        view.findViewById(R.id.btn_save_changes).setOnClickListener(v -> onClickSave());
        return view;

    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        etFirstName.setText(firstName);
        etLastName.setText(lastName);
        etBirthdate.setText(DateUtil.parseDate(dob, PATTERN_RFC1123, "MM/dd/YYYY"));
        etBirthdate.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                if (b) {
                    openDatePicker();
                }
            }
        });
        etBirthdate.setOnClickListener(v -> openDatePicker());
        etEmail.setText(email);
        if (isMale) {
            rbMale.setChecked(true);
        } else {
            rbFemale.setChecked(true);
        }

        editClientViewModel.getDataPatientsLive().observe(this.getViewLifecycleOwner(), viewStateObserver);

    }

    Observer<HomeClientsViewState> viewStateObserver = viewState -> {
        if (viewState instanceof HomeClientsViewState.PatientSuccess) {
            dismiss();
            listener.onClientUpdated();
        }
    };

    private void openDatePicker(){
        Date date = DateUtil.parseDate(dob, DateUtil.PATTERN_RFC1123);
        MaterialDatePicker datePicker = MaterialDatePicker.Builder.datePicker()
                .setSelection(date != null ? date.getTime() : MaterialDatePicker.todayInUtcMilliseconds())
                .setTitleText("Select Birth date").build();

        datePicker.show(getChildFragmentManager(), datePicker.toString());
        datePicker.addOnPositiveButtonClickListener(selection -> {
            String dateString = DateUtil.formatDate((new Date((Long) selection)), "MM/dd/YYYY");
            etBirthdate.setText(dateString);
        });
    }


    @Override
    public int getTheme() {
        return R.style.CustomBottomSheetDialog;
    }

    public void onClickSave() {
        editClientViewModel.updateClient(id,
                etFirstName.getText().toString(),
                etLastName.getText().toString(),
                DateUtil.parseDate(etBirthdate.getText().toString(), "MM/dd/yyyy", PATTERN_RFC1123),
                rbMale.isChecked(),
                etEmail.getText().toString()
                );
    }
}

File: app/src/main/java/com/waveneuro/ui/dashboard/home/bottom_sheet/edit_client/EditClientBottomSheet.kt
package com.waveneuro.ui.dashboard.home.bottom_sheet.edit_client

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.datepicker.MaterialDatePicker
import com.waveneuro.R
import com.waveneuro.domain.model.common.SexType
import com.waveneuro.databinding.DialogEditClientBinding
import com.waveneuro.ui.base.utils.EventObserver
import com.waveneuro.ui.dashboard.home.bottom_sheet.edit_client.EditClientViewState.Success
import com.waveneuro.utils.DateHelper
import com.waveneuro.utils.ext.toast
import java.util.*

class EditClientBottomSheet : BottomSheetDialogFragment() {

    private val viewModel: EditClientViewModel by viewModels()

    private lateinit var binding: DialogEditClientBinding

    private var clientId = 0
    private var firstName: String? = null
    private var lastName: String? = null
    private var birthday: String? = null
    private var isMale: Boolean = false
    private var onClientUpdated: ((String?) -> Unit)? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DialogEditClientBinding.inflate(layoutInflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        with(binding) {
            tipFirstName.setText(firstName ?: "")
            tipLastName.setText(lastName ?: "")
            tipBirthday.setText(DateHelper.birthdayFormat(birthday))

            rbMale.isChecked = isMale
            rbFemale.isChecked = !isMale

            btnSaveChanges.setOnClickListener { onClickSave() }
            tipBirthday.setOnClickListener { showDatePicker() }
        }

        setObserver()
    }

    private fun setObserver() {
        viewModel.dataEditClientLive.observe(requireActivity(), Observer { viewState ->
            when (viewState) {
                is Success -> {
                    onClientUpdated?.invoke(viewState.fullName)
                    dismiss()
                }
            }
        })
        viewModel.message.observe(viewLifecycleOwner, EventObserver(::showMessage))
    }

    private fun showMessage(message: String?) {
        requireActivity().toast(message)
    }

    private fun showDatePicker() {
        MaterialDatePicker.Builder.datePicker()
            .setSelection(DateHelper.getBirthdayDate(birthday)?.time)
            .setTitleText(R.string.date_picker_birth_title)
            .build().apply {
                addOnPositiveButtonClickListener {
                    birthday = DateHelper.birthdayFormat(Date(it))
                    binding.tipBirthday.setText(birthday ?: "")
                }
            }
            .show(childFragmentManager, "DatePicker")
    }

    override fun getTheme(): Int {
        return R.style.CustomBottomSheetDialog
    }

    private fun onClickSave() {
        with(binding) {
            viewModel.updateClient(
                clientId,
                tipFirstName.text?.trim().toString(),
                tipLastName.text?.trim().toString(),
                tipBirthday.text?.trim().toString(),
                if (rbMale.isChecked) SexType.MALE else SexType.FEMALE
            )
        }
    }

    companion object {
        fun newInstance(
            onClientUpdated: ((String?) -> Unit)?,
            id: Int,
            name: String?,
            lastName: String?,
            birthday: String?,
            sex: Boolean,
        ) = EditClientBottomSheet().apply {
            this.clientId = id
            this.firstName = name
            this.lastName = lastName
            this.birthday = birthday
            this.isMale = sex
            this.onClientUpdated = onClientUpdated
        }
    }

}